pkg/cmd/auth/shared/ssh_keys.go

```diff
@@ -0,0 +1,153 @@
+package shared
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+
+	"github.com/AlecAivazis/survey/v2"
+	"github.com/cli/cli/api"
+	"github.com/cli/cli/internal/config"
+	"github.com/cli/cli/internal/ghinstance"
+	"github.com/cli/cli/internal/run"
+	"github.com/cli/cli/pkg/prompt"
+	"github.com/cli/safeexec"
+)
+
+func localPublicKeys() ([]string, error) {
+	sshDir, err := config.HomeDirPath(".ssh")
+	if err != nil {
+		return nil, err
+	}
+
+	return filepath.Glob(filepath.Join(sshDir, "*.pub"))
+}
+
+func findKeygen() (string, error) {
+	keygenExe, err := safeexec.LookPath("ssh-keygen")
+	if err != nil && runtime.GOOS == "windows" {
+		// We can try and find ssh-keygen in a Git for Windows install
+		if gitPath, err := safeexec.LookPath("git"); err == nil {
+			gitKeygen := filepath.Join(filepath.Dir(gitPath), "..", "usr", "bin", "ssh-keygen.exe")
+			if _, err = os.Stat(gitKeygen); err == nil {
+				return gitKeygen, nil
+			}
+		}
+	}
+	return keygenExe, err
+}
+
+func generateSSHKey() (string, error) {
+	keygenExe, err := findKeygen()
+	if err != nil {
+		// give up silently if `ssh-keygen` is not available
+		return "", nil
+	}
+
+	var sshChoice bool
+	err = prompt.SurveyAskOne(&survey.Confirm{
+		Message: "Generate a new SSH key to add to your GitHub account?",
+		Default: true,
+	}, &sshChoice)
+	if err != nil {
+		return "", fmt.Errorf("could not prompt: %w", err)
+	}
+	if !sshChoice {
+		return "", nil
+	}
+
+	sshDir, err := config.HomeDirPath(".ssh")
+	if err != nil {
+		return "", err
+	}
+	keyFile := filepath.Join(sshDir, "id_ed25519")
+	if _, err := os.Stat(keyFile); err == nil {
+		return "", fmt.Errorf("refusing to overwrite file %s", keyFile)
+	}
+
+	if err := os.MkdirAll(filepath.Dir(keyFile), 0711); err != nil {
+		return "", err
+	}
+
+	var sshLabel string
+	// err = prompt.SurveyAskOne(&survey.Input{
+	// 	Message: "Enter the label for your new SSH key",
+	// }, &sshLabel)
+	// if err != nil {
+	// 	return "", fmt.Errorf("could not prompt: %w", err)
+	// }
+
+	var sshPassphrase string
+	err = prompt.SurveyAskOne(&survey.Password{
+		Message: "Enter the passphrase for your new SSH key",
```

Comment:
I think this ought to be called out as optional.

Reply:
```suggestion
		Message: "Enter a passphrase for your new SSH key (Optional)",
```

What @vilmibm said, and also should we say "a passphase" instead of "the passphrase"? If I saw "the" I would expect that there was an existing passphrase I should know about, and this is just entering a new one, right?

Reply:
Great call about the indefinite article, thank you!